common/numerated/src/tree.rs

```diff
@@ -0,0 +1,835 @@
+use scale_info::{
+    scale::{Decode, Encode},
+    TypeInfo,
+};
+
+/// # Non overlapping intervals tree.
+/// Can be considered as set of points, but with possibility to work with
+/// continuous sets of this points (the same as interval) as fast as with points.
+/// Insert and remove operations are O(log(n)) complexity, where n is amount of intervals.
+/// So, even if you insert for example points from 0u64 to u64::MAX,
+/// then removing all of them or any part of them is as fast as removing one point.
+///
+/// # Functionality examples:
+/// ```
+/// use numerated::IntervalsTree;
+/// use std::collections::BTreeSet;
+/// use std::ops::RangeInclusive;
+///
+/// let mut tree = IntervalsTree::new();
+/// let mut set = BTreeSet::new();
+///
+/// tree.insert(1i32);
+/// // now `tree` contains only one interval: [1..=1]
+/// set.insert(1i32);
+/// // `points_iter()` - is iterator over all points in `tree`.
+/// assert_eq!(set, tree.points_iter().collect());
+///
+/// // We can add points from 3 to 100_000 only by one insert operation.
+/// // `try` is only for range check, that it has start <= end.
+/// tree.try_insert(3..=100_000).unwrap();
+/// // now `tree` contains two intervals: [1..=1] and [3..=100_000]
+/// set.extend(3..=100_000);
+/// // extend complexity is O(n), where n is amount of elements in range.
+/// assert_eq!(set, tree.points_iter().collect());
+///
+/// // We can remove points from 1 to 99_000 not inclusive only by one remove operation.
+/// // `try` is only for range check, that it has start <= end.
+/// tree.try_remove(1..99_000).unwrap();
+/// // now `tree` contains two intervals: [99_000..=100_000]
+/// (1..99_000).for_each(|i| { set.remove(&i); });
+/// // remove complexity for set is O(n*log(m)),
+/// // where `n` is amount of elements in range and `m` size of set.
+/// assert_eq!(set, tree.points_iter().collect());
+///
+/// // Can insert or remove all possible points just by one operation:
+/// tree.insert(..);
+/// tree.remove(..);
+///
+/// // Iterate over voids (intervals between intervals in tree):
+/// tree.try_insert(1..=3).unwrap();
+/// tree.try_insert(5..=7).unwrap();
+/// let voids = tree.voids(..).map(RangeInclusive::from).collect::<Vec<_>>();
+/// assert_eq!(voids, vec![i32::MIN..=0, 4..=4, 8..=i32::MAX]);
+///
+/// // AndNot iterator: iterate over intervals from `tree` which are not in `other_tree`.
+/// let other_tree: IntervalsTree<i32> = [3, 4, 5, 7, 8, 9].iter().collect();
+/// let and_not: Vec<_> = tree.and_not_iter(&other_tree).map(RangeInclusive::from).collect();
+/// assert_eq!(and_not, vec![1..=2, 6..=6]);
+/// ```
+///
+/// # Possible panic cases.
+/// Using `IntervalsTree` for type `T: Numerated` cannot cause panics,
+/// if implementation `Numerated`, `Copy`, `Ord`, `Eq` are correct for `T`.
+/// In other cases `IntervalsTree` does not guarantees execution without panics.
+#[derive(Clone, PartialEq, Eq, TypeInfo, Encode, Decode)]
+pub struct IntervalsTree<T> {
```

Comment:
```rust
impl<T> Deref<BTreeMap<T, T>> for IntervalsTree<T> {}
impl<T> DerefMut<BTreeMap<T, T>> for IntervalsTree<T> {}
```

Reply:
No, this is not supposed to be in access for user